grails-app/services/org/pih/warehouse/order/OrderService.groovy

```diff
@@ -534,20 +537,72 @@ class OrderService {
+                        if (orderItem.order.id != orderId) {
+                            throw new UnsupportedOperationException("You can not edit items from another order!")
+                        }
+                    } else {
+                        orderItem = new OrderItem()
+                    }
 
+                    Product product
                     if (productCode) {
-                        def product = Product.findByProductCode(productCode)
+                        product = Product.findByProductCode(productCode)
                         if (!product) {
                             throw new ProductException("Unable to locate product with product code ${productCode}")
                         }
-                        OrderItem orderItem = new OrderItem(product: product, quantity: quantity, unitPrice: unitPrice)
-                        order.addToOrderItems(orderItem)
-                        count++
+                        orderItem.product = product
+                    }
+
+                    if (sourceCode) {
+                        orderItem.productSupplier = ProductSupplier.findByCode(sourceCode)
+                    } else if (supplierCode && manufacturer && manufacturerCode) {
+                        if (Organization.get(manufacturer)) {
+                            Organization supplier = Organization.get(supplierId)
+                            def supplierParams = [manufacturer: manufacturer,
+                                                  product: product,
+                                                  supplierCode: supplierCode,
+                                                  manufacturerCode: manufacturerCode,
+                                                  supplier: supplier]
+                            ProductSupplier productSupplier = productSupplierDataService.getOrCreateNew(supplierParams)
+                            orderItem.productSupplier = productSupplier
+                        }
                     }
 
+                    if (unitOfMeasure) {
+                        String[] uomParts = unitOfMeasure.split("/")
+                        if (uomParts.length <= 1 || !UnitOfMeasure.findByName(uomParts[0])) {
+                            throw new IllegalArgumentException("Could not find provided Unit of Measure: ${unitOfMeasure}.")
+                        }
+                        UnitOfMeasure uom = uomParts.length > 1 ? UnitOfMeasure.findByName(uomParts[0]) : null
+                        BigDecimal qtyPerUom = uomParts.length > 1 ? BigDecimal.valueOf(Double.valueOf(uomParts[1])) : null
+                        orderItem.quantityUom = uom
+                        orderItem.quantityPerUom = qtyPerUom
+                    }
+
+                    orderItem.quantity = Integer.valueOf(quantity)
+                    orderItem.unitPrice = BigDecimal.valueOf(Integer.valueOf(unitPrice))
+                    orderItem.recipient = recipientNames.length > 1 ?
```

Comment:
Consider reusing/improving PersonDataService.createOrUpdatePerson(Map)

Reply:
In fact, search the code by "findByFirstNameAndLastName" and there should be three references to similar code. Let's try to refactor it so we're reusing the same code in all three places (if it makes sense).

Reply:
Ok, will do

Reply:
Hmm, I just noticed, that in requirements from this ticket there is said that if the person is not found, it should not create a new one. The code you mentioned is creating a new Person if they can't find it by first and last name. I'll just refactor the code using findByFirstNameAndLastName.